Assets/MRTK/SDK/Experimental/Drawstring/Scripts/ElasticSystem.cs

```diff
@@ -0,0 +1,196 @@
+        public T[] snapPoints;
+    }
+
+    /// <summary>
+    /// Properties of the damped harmonic oscillator differential system.
+    /// </summary>
+    [Serializable]
+    public struct ElasticProperties
+    {
+        /// <value>
+        /// Mass of the simulated oscillator element
+        /// </value>
+        [SerializeField]
+        public float mass;
+        /// <value>
+        /// Hand spring constant
+        /// </value>
+        [SerializeField]
+        public float hand_k;
+        /// <value>
+        /// End cap spring constant
+        /// </value>
+        [SerializeField]
+        public float end_k;
+        /// <value>
+        /// Snap point spring constant
+        /// </value>
+        [SerializeField]
+        public float snap_k;
+        /// <value>
+        /// Extent at which snap points begin forcing the spring.
+        /// </value>
+        [SerializeField]
+        public float snap_radius;
+        /// <value>
+        /// Drag/damper factor, proportional to velocity.
+        /// </value>
+        [SerializeField]
+        public float drag;
+    }
+
+    /// <summary>
+    /// Represents a damped harmonic oscillator over an
+    /// N-dimensional vector space, specified by generic type T.
+    /// 
+    /// This extensibility allows not just for 1, 2, and 3-D springs, but
+    /// allows for 4-dimensional quaternion springs.
+    /// </summary>
+    internal abstract class ElasticSystem<T>
+    {
+        protected ElasticExtentProperties<T> extentInfo;
+        protected ElasticProperties elasticProperties;
+
+        protected T currentValue;
+        protected T currentVelocity;
+
+        public ElasticSystem(T initialValue, T initialVelocity,
+                                ElasticExtentProperties<T> extentInfo, ElasticProperties elasticProperties)
+        {
+            this.extentInfo = extentInfo;
+            this.elasticProperties = elasticProperties;
+            currentValue = initialValue;
+            currentVelocity = initialVelocity;
+        }
+
+        /// <summary>
+        /// Update the internal state of the damped harmonic oscillator, given the forcing/desired value.
+        /// </summary>
+        /// <param name="forcingValue">Input value, for example, a desired manipulation position.</param>
+        /// <param name="deltaTime">Amount of time that has passed since the last update.</param>
+        public abstract T ComputeIteration(T forcingValue, float deltaTime);
+
+        /// <summary>
+        /// Query the elastic system for the current instantaneous value
+        /// </summary>
+        /// <returns>Current value of the elastic system</returns>
+        public T GetCurrentValue() => currentValue;
+    }
+
+    internal class LinearElasticSystem : ElasticSystem<float>
+    {
+        public LinearElasticSystem(float initialValue, float initialVelocity,
+                                ElasticExtentProperties<float> extentInfo, ElasticProperties elasticProperties)
```

Comment:
Nit: Fix indentation here and on the next line

Reply:
Fixed